### PR TITLE
fix: add more tests to deserialization

### PR DESCRIPTION
**Pull request by mateusoliveira43**
Signed-off-by: Mateus Oliveira <msouzaol@redhat.com>

## Describe the behavior changes introduced in this PR

Add more tests to deserialization module and also running doctests when running the test suite.

## Linked Issues

Resolves https://github.com/konveyor/pelorus/issues/754

## Testing Instructions

run test and coverage command 
```
coverage run -m pytest -rap -m "not integration and not mockoon" && coverage html
```
to check tests and coverage.

### CODE REVIEW

File: exporters/tests/test_deserialization.py [hunk 1]
```diff
@@ -50,10 +115,11 @@ class Missing:
     print(e.value)
 
 
-def test_nested_field_positive():
+@pytest.mark.parametrize("nested_path", ["foo.bar", ("foo", "bar")])
```
Comment (KevinMGranger) [inline exporters/tests/test_deserialization.py:118]: :+1:

File: exporters/pelorus/deserialization/__init__.py [hunk 2]
```diff
@@ -223,9 +217,6 @@ def _extract_optional_type(type_: type) -> Optional[type]:
     """
     If this type annotation is Optional[T], then returns the type T.
     Otherwise, returns None.
-
-    >>> assert _extract_optional_type(Optional[int]) == int
-    >>> assert _extract_optional_type(int) is None
```
Comment (KevinMGranger) [inline exporters/pelorus/deserialization/__init__.py:228]: I don't think we should remove the doctests. They're just as much documentation as they are tests.

What we can look into later is pytest's doctest support: https://docs.pytest.org/en/7.1.x/how-to/doctest.html

File: exporters/tests/test_deserialization.py [hunk 3]
```diff
@@ -9,12 +9,77 @@
...
     nested,
     retain_source,
 )
 
 
+@pytest.mark.parametrize(
+    "type_",
+    [list[int], str, set[str], Iterable[str], Container[str], tuple[float, float]],
+)  # Iterable and container should not have __getitem__, tuple should pass?
+def test_extract_dict_types_from_non_dict(type_: type):
```
Comment (KevinMGranger) [inline exporters/tests/test_deserialization.py:25]: I don't think we should test implementation details here. This should be tested as part of the regular testing.

There's this idea of "the testing pyramid", where:
-  integration tests are large, harder to write (and sometimes maintain), but very valuable even for one test.
-  unit tests are small, simple, and easy to maintain, but less valuable individually.

The deserialization unit tests for classes are "bigger" unit tests, and will hit this code path regardless.

File: exporters/tests/test_deserialization.py [hunk 4]
```diff
@@ -9,12 +9,77 @@
...
     deserialize,
     nested,
     retain_source,
 )
 
 
+@pytest.mark.parametrize(
+    "type_",
+    [list[int], str, set[str], Iterable[str], Container[str], tuple[float, float]],
+)  # Iterable and container should not have __getitem__, tuple should pass?
```
Comment (KevinMGranger) [inline exporters/tests/test_deserialization.py:24]: I don't know if it's worth it to catch every edge case. This was mainly for working with openshift objects, and I don't know when we'd actually expect a tuple!

**Review by KevinMGranger — CHANGES_REQUESTED** [review]
I like the nested and Any tests you added! But let's discuss or implement my notes about the doctests and what kind of unit tests we should write.

File: exporters/pelorus/deserialization/__init__.py (hunk 2 above)
Comment (mateusoliveira43) [reply exporters/pelorus/deserialization/__init__.py:228]: I look into it

File: exporters/tests/test_deserialization.py (hunk 4 above)
Comment (mateusoliveira43) [reply exporters/tests/test_deserialization.py:24]: I would add a comment about not supporting tuples then

File: exporters/tests/test_deserialization.py (hunk 3 above)
Comment (mateusoliveira43) [reply exporters/tests/test_deserialization.py:25]: this one I was trying to hit one of the ifs of the function (And I was only able with set), but I will remove the more unusual types

File: exporters/pelorus/utils/nested.py [hunk 5]
```diff
@@ -151,21 +155,21 @@ class BadAttributePathError(Exception):
...
     path: Sequence[str]
     path_slice: slice
+    key: str
     value: Any
     root_name: Optional[str] = None
 
     @property
     def message(self) -> str:
-        msg = f"{self.root_name + ' is missing' if self.root_name else 'Missing'} {'.'.join(self.path)}"
+        msg = f"{self.root_name + ' is missing' if self.root_name else 'Missing'} {self.key}"
```
Comment (mateusoliveira43) [inline exporters/pelorus/utils/nested.py:166]: I changed here because the errors message where not meaningful for me

THEN
```
exporters/tests/test_deserialization.py::test_nested_missing Error while deserializing <class 'tests.test_deserialization.test_nested_missing.<locals>.Nested'> from dict:
  nested_int: Missing foo.bar foo.bar because foo was {}


exporters/tests/test_deserialization.py::test_multi_nested_missing Error while deserializing <class 'tests.test_deserialization.test_multi_nested_missing.<locals>.Nested'> from dict:
  nested_int: Missing foo.com.example.int foo.com.example.int because foo was {}
```
NOW
```
exporters/tests/test_deserialization.py::test_nested_missing Error while deserializing <class 'tests.test_deserialization.test_nested_missing.<locals>.Nested'> from dict:
  nested_int: Missing bar in foo.bar because foo was {}

exporters/tests/test_deserialization.py::test_multi_nested_missing Error while deserializing <class 'tests.test_deserialization.test_multi_nested_missing.<locals>.Nested'> from dict:
  nested_int: Missing com in foo.com[example.int] because foo was {}
```

File: exporters/pelorus/deserialization/__init__.py (hunk 2 above)
Comment (mateusoliveira43) [reply exporters/pelorus/deserialization/__init__.py]: added it :)

**Comment by KevinMGranger** [comment]
I'm going to revisit the unit tests part and clean it up myself (perhaps even cleaning up the code as we go along).

/lgtm

**Review by KevinMGranger — APPROVED** [review]
